sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,919 @@
+
+    public HadoopInputFormatBoundedSource(SerializableConfiguration conf, Coder<K> keyCoder,
+        Coder<V> valueCoder) {
+      this(conf, keyCoder, valueCoder, null, null, null);
+    }
+
+    public HadoopInputFormatBoundedSource(SerializableConfiguration conf, Coder<K> keyCoder,
+        Coder<V> valueCoder, SimpleFunction<?, K> keyTranslation,
+        SimpleFunction<?, V> valueTranslation, SerializableSplit serializableSplit) {
+      this.conf = conf;
+      this.serializableSplit = serializableSplit;
+      this.keyCoder = keyCoder;
+      this.valueCoder = valueCoder;
+      this.simpleFuncForKeyTranslation = keyTranslation;
+      this.simpleFuncForValueTranslation = valueTranslation;
+    }
+
+    public SerializableConfiguration getConfiguration() {
+      return conf;
+    }
+
+    @Override
+    public void validate() {
+      checkNotNull(conf, "Need to set the configuration of a HadoopInputFormatSource");
+      checkNotNull(keyCoder, "KeyCoder should not be null in HadoopInputFormatSource");
+      checkNotNull(valueCoder, "ValueCoder should not be null in HadoopInputFormatSource");
+    }
+
+    // Indicates if the source is split or not
+    private boolean isSourceSplit = false;
+
+    public boolean isSourceSplit() {
+      return isSourceSplit;
+    }
+
+    public void setSourceSplit(boolean isSourceSplit) {
+      this.isSourceSplit = isSourceSplit;
+    }
+
+    @Override
+    public List<BoundedSource<KV<K, V>>> splitIntoBundles(long desiredBundleSizeBytes,
+        PipelineOptions options) throws Exception {
+      if (serializableSplit == null) {
+        if (inputSplits == null) {
+          computeSplits();
+        }
+        LOG.info("Generated {} splits each of size {} ", inputSplits.size(),
+            inputSplits.get(0).getSplit().getLength());
+        setSourceSplit(true);
+        return Lists.transform(inputSplits,
+            new Function<SerializableSplit, BoundedSource<KV<K, V>>>() {
+              @Override
+              public BoundedSource<KV<K, V>> apply(SerializableSplit serializableInputSplit) {
+                HadoopInputFormatBoundedSource<K, V> hifBoundedSource =
+                    new HadoopInputFormatBoundedSource<K, V>(conf, keyCoder, valueCoder,
+                        simpleFuncForKeyTranslation, simpleFuncForValueTranslation,
+                        serializableInputSplit);
+                return hifBoundedSource;
+              }
+            });
+      } else {
+        setSourceSplit(true);
+        LOG.info("Not splitting source {} because source is already split.", this);
+        return ImmutableList.of((BoundedSource<KV<K, V>>) this);
+      }
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions po) throws Exception {
+      if (inputSplits == null) {
+        computeSplits();
+      }
+      return boundedSourceEstimatedSize;
+    }
+
+    private long boundedSourceEstimatedSize = 0;
+    private List<SerializableSplit> inputSplits;
```

Comment:
inputSplits needs to be marked transient - otherwise it could get serialized. It's really just a cache holding onto the splits between when they're grabbed in getEstimatedSizeBytes until they're used in computeSplits